### ISSUE TITLE
Support for Visual Studio 2017

### ISSUE THREAD

**Issue by Ascrod** [MoonchildProductions/UXP#1213]
There was an issue for investigating this on the old PM repo: https://github.com/MoonchildProductions/Pale-Moon/issues/902. I presume it was closed with the intention of revisiting the question after the UXP migration.

Since we're now on UXP and VS 2015, is it still be desirable to move to or support building with VS 2017 on Windows?

**Comment by mattatobin** [comment]
There wouldn't be much benefit and we do have other things to do AND this is not in any way trivial.

**Comment by wolfbeast** [comment]
The benefit there would be having the advantage of some more C++11/C++14 language features available.
It's not trivial, but also not excessively difficult -- but it would mean we'd need **very** intensive testing of the result since switching compiler is not unlike touching every line of code.

**Comment by Ascrod** [comment]
For this, I think we can start with [Bug 1318143](https://bugzilla.mozilla.org/show_bug.cgi?id=vs2017) and its dependencies where applicable. I'd also like to add [Bug 1356493](https://bugzilla.mozilla.org/show_bug.cgi?id=1356493), which adds support for the Windows 10 SDK, because the SDK installs by default with VS2017 and the patch is pretty trivial.

I think these patches provide backwards compatibility with VS2015, but it's worth testing to be sure since I assume it's something we want to do.

Mind if I take a stab at this?

**Comment by wolfbeast** [comment]
> Mind if I take a stab at this?

Go ahead.

**Comment by wolfbeast** [comment]
I think it might be a good idea to add the Win10 SDK support regardless of the VS2017 question for optimal compat with the current windows version. so, best split that off into its own issue and tackle that first?

**Comment by Ascrod** [comment]
Can do.

**Comment by Ascrod** [comment]
A few of the patches I'm looking at implementing either move some existing functionality out of old-configure (autoconf) into moz.configure (python) or just add new functionality into moz.configure. I seem to recall hearing from somewhere a while back that this isn't desirable for us in the long term... is that the case? Does it matter?

Examples are [Bug 1371465](https://bugzilla.mozilla.org/show_bug.cgi?id=1371465) the second patch from [Bug 1318143](https://bugzilla.mozilla.org/show_bug.cgi?id=1318143).

**Comment by mattatobin** [comment]
Yes it does matter.

**Comment by Ascrod** [comment]
> 
> 
> Yes it does matter.

Details, please? Should I be adding as much as I can to old-configure? moz.configure is already being used in places, so *some* changes to it will be unavoidable, but I can try to minimize them if that is what you are after.

**Comment by wolfbeast** [comment]
You should be adding things exclusively to old-configure. If you don't, then we'll have to revert those changes again later to convert it to old-configure which is an on-going process. Let's avoid that double work, please.

**Comment by Ascrod** [comment]
As an example of what I'm talking about, [this is one of the changesets I'm looking at.](https://hg.mozilla.org/mozilla-central/rev/311336b5377d) This and other changesets related to supporting VS2017 involve a non-trivial amount of changes to moz.config; I can't tell how easy it will be to move things like the `vc_compiler_path` function, and everything that depends upon it, out of moz.config and into old-configure, without going down a rabbit hole that will likely just lead to migrating the rest of moz.configure anyway.

I'll go ahead and post all of the bugs I know of that at least need to be looked at, to give you a rough idea of the scope:

- [Bug 1318143](https://bugzilla.mozilla.org/show_bug.cgi?id=1318143) - Detect VS2017, add option to use VS2015
- [Bug 1371465](https://bugzilla.mozilla.org/show_bug.cgi?id=1371465) - Set MSVS_VERSION for VS2017
- [Bug 1484184](https://bugzilla.mozilla.org/show_bug.cgi?id=1484184) - Fix _RAISE macro error
- [Bug 1522560](https://bugzilla.mozilla.org/show_bug.cgi?id=1522560) - Use installed vswhere binary
- [Bug 1362872](https://bugzilla.mozilla.org/show_bug.cgi?id=1362872) - Use mbcs encoding for vswhere output
- [Bug 1318182](https://bugzilla.mozilla.org/show_bug.cgi?id=1318182) - Generate VS2017 project files
- [Bug 1395907](https://bugzilla.mozilla.org/show_bug.cgi?id=1395907) - Require VS2017u3+
- [Bug 1424281](https://bugzilla.mozilla.org/show_bug.cgi?id=1424281) - Require VS2017u6+
(With regards to these last two, I assume we're only interested in supporting VS2017u9 and later.)
- [Bug 1360631](https://bugzilla.mozilla.org/show_bug.cgi?id=1360631) - Fix VS2015 x86 cross compiling
- [Bug 1361750](https://bugzilla.mozilla.org/show_bug.cgi?id=1361750) - Disable some VS2017 warnings in PSM
- [Bug 1411600](https://bugzilla.mozilla.org/show_bug.cgi?id=1411600) - Add Add VS2017 mozconfigs
Various compile error fixes:
- [Bug 1310067](https://bugzilla.mozilla.org/show_bug.cgi?id=1310067)
- [Bug 1319971](https://bugzilla.mozilla.org/show_bug.cgi?id=1319971)
- [Bug 1345331](https://bugzilla.mozilla.org/show_bug.cgi?id=1345331)
- [Bug 1358169](https://bugzilla.mozilla.org/show_bug.cgi?id=1358169)
- [Bug 1408695](https://bugzilla.mozilla.org/show_bug.cgi?id=1408695)
- [Bug 1423649](https://bugzilla.mozilla.org/show_bug.cgi?id=1423649)
- [Bug 1443367](https://bugzilla.mozilla.org/show_bug.cgi?id=1443367)

**Comment by wolfbeast** [comment]
That's quite the scope. Looking at all that I'm not sure if we actually want that without a clear and present advantage of switching to VS2017 over what we have now. Do any of the bugs give a clear indication of something that is crucial to have that is in 2017 and not in 2015?

**Comment by mattatobin** [comment]
I don't believe that info is in the bugs even if it was the case considering the pre-clang attitude was newer is always better. 

I do know it hasn't done SeaMonkey any favors and my vauge impression at the time was that Firefox suffered both after the switch and when they failed to make a transition to 2019.

We have years yet before 2015 EoLs in yes.. 2025. VS2017 just isn't worth it in my opinion and who the hell knows if we will even have a Microsoft Windows to speak of by then.

**Comment by Ascrod** [comment]
VS2017 should be pretty mature by now, so we likely won't run into as many issues with it as Mozilla did. They started at supporting the pre-release and kept having to work around compiler issues in MSVC that weren't fixed until later updates, or changes that were introduced in later versions.

These changes *do* include a new switch to allow continuing to use 2015 if it's installed, so it won't be forcing anyone to change their dev environment right away. But it will give new developers an option to use 2017 if they have it rather than having to break *their* environment just to install 2015.

Lastly, yeah VS2015 will be in support until 2025, but we'll have to do this upgrade then anyway, yes? 😉

I will continue to experiment with this, with your blessing. Like I said before, I can keep as much of this as I can in autoconf, but some changes to moz.config will be unavoidable without redoing that part of the build system anyway.

**Comment by mattatobin** [comment]
We do NOT have to do this. That is nonsense. And 2017 would give us only 2 more years support wise AND 2019 is impossible without huge swaths of rewrites and at that point it is Visual Studio as a Service.

**Comment by wolfbeast** [comment]
It's fine to change things in moz.configure if you can weave all of this in a way that doesn't influence vs2015 use. We can evaluate later what is best to use for releases, and we can also deal with the autoconf stuff later.

I agree with @mattatobin that there's no "newer is always better" driver here. By that mode we should have abandoned MSVC for the more rapidly released clang/gcc already but those are simply **not** the best tools for the job here.

But my question is still: what does 2017 bring to the table that we "do want"? I've examined the abstract concepts introduced in c++11, c++14 and c++17 and to me it seems a lot of that is stuff we actually don't need (or want). C++ is already too malleable and it'd only promote bad coding, IMHO.

**Comment by Ascrod** [comment]
> But my question is still: what does 2017 bring to the table that we "do want"? I've examined the abstract concepts introduced in c++11, c++14 and c++17 and to me it seems a lot of that is stuff we actually don't need (or want). C++ is already too malleable and it'd only promote bad coding, IMHO.

I'm afraid I don't know enough about the different C++ versions to really answer this. I merely saw this issue on the old repo, assumed it was beneficial, and thought it got lost in translation to UXP.

> But like I said, continue your research project but know it may never see the light of day if it doesn't pan out or prove superior.

I am willing to put in the effort, but I readily admit that I am not experienced enough to know every problem that might arise or how to solve it without support from others who know this stuff better. I don't really have any personal stake in supporting newer versions of VS; I just thought it might be a useful feature to contribute. But if it's not desired by the core development team, then it doesn't have to be implemented. That's your call.

**Comment by mattatobin** [comment]
Well we can't know until it builds. If we are to switch or stay we should have a reason to combat either "older version was better" AND the "old and obsolete" crowds.

This is why I have shifted to supporting the research. But I remain cautions.
